{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "805a75db",
   "metadata": {},
   "source": [
    "# Population genetics\n",
    "\n",
    "Let's do some basic population genetics with what we have at hand here\n",
    "\n",
    "```\n",
    "module load bcftools R\n",
    "\n",
    "TGDATA=\"/lisc/scratch/admixlab/share/vds_course/ALL.chr18.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz\" \n",
    "```\n",
    "\n",
    "\n",
    "## Heterozygosity\n",
    "\n",
    "Now, we may go back to the 1000 genomes VCF file we used earlier, and obtain heterozygous positions. For this, we can use `bcftools`. With `-s` samples are selected, and the rows starting on `PSC` are per-sample-counts. This table could be the raw table.\n",
    "\n",
    "\n",
    "```\n",
    "bcftools stats -s - $TGDATA | grep \"^PSC\" -B 1 > hets.txt\n",
    "```\n",
    "\n",
    "Keep in mind that, this as well as many other things can be solved by multiple possible strategies. In this case, I show one version, but there are alternatives.\n",
    "\n",
    "If you inspect the file, you will see that there is one line per sample, with several columns containing the statistics. Column 6 is the number of Hets.\n",
    "\n",
    "* Let's have some look at the data!\n",
    "\n",
    "```\n",
    "# in R\n",
    "\n",
    "R --vanilla\n",
    "\n",
    "hets<-read.table(\"hets.txt\", sep=\"\\t\",header=T,comment.char=\"\")\n",
    "\n",
    "\n",
    "png(\"hets_histogram.png\",600,600)\n",
    "hist(hets[,6])\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "* Is there a relationship between homozygous alternative (1/1) and heterozygous (0/1) sites?\n",
    "\n",
    "```\n",
    "png(\"hets_vs_homALT.png\",600,600)\n",
    "plot(hets[,5],hets[,6])\n",
    "dev.off()\n",
    "\n",
    "cor.test(hets[,5],hets[,6])\n",
    "```\n",
    "\n",
    "* Now, I would like to stratify this by population groups. For this, we need to load the metadata and do some merging. Unfortunately, when creating this file, the authors messed with empty columns, so we need to tell `R` to ignore this. Then we merge, taking only the sample ID and heterozygous call columns.\n",
    "\n",
    "\n",
    "```\n",
    "meta<-read.table(\"/lisc/scratch/admixlab/share/vds_course/integrated_call_samples_v3.20130502.ALL.panel\", sep=\"\\t\",header=T,fill=T)\n",
    "\n",
    "head(meta)\n",
    "\n",
    "hetmet<-merge(hets[,c(3,6)],meta,by.x=1,by.y=1)\n",
    "\n",
    "head(hetmet)\n",
    "```\n",
    "\n",
    "* Just having some fun with statistics in R:\n",
    "\n",
    "```\n",
    "mean(hetmet[which(hetmet$pop==\"IBS\"),2])\n",
    "mean(hetmet[which(hetmet$pop==\"FIN\"),2])\n",
    "wilcox.test(hetmet[which(hetmet$pop==\"IBS\"),2],hetmet[which(hetmet$pop==\"FIN\"),2])\n",
    "```\n",
    "\n",
    "* Then, let's make a boxplot stratified by continental population (already adding colour and labels to nice it up):\n",
    "\n",
    "```\n",
    "png(\"hets_superpop_boxplot.png\",600,600)\n",
    "boxplot(X.6.nHets~super_pop,data=hetmet,\n",
    "    col=c(\"blue\",\"green\",\"orange\",\"yellow\",\"red\"),\n",
    "    xlab=\"Continental population\",ylab=\"Heterozygous sites\")\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "This very simple statistic does have a biological meaning, hence it is very informative!\n",
    "\n",
    "* A final thing: nicely ordering the data by super-population, and plotting the distribution by more specific populations. This requires a bit of `R`-specific data handling, to get a properly sorted dataframe.\n",
    "\n",
    "```\n",
    "mypops<-c(\"AFR\",\"SAS\",\"EUR\",\"EAS\",\"AMR\")\n",
    "sortedpops<-list()\n",
    "for (npop in mypops) { sortedpops[[npop]]<-unique(hetmet$pop[which(hetmet$super_pop==npop)]) }\n",
    "hetmetnice<-data.frame(nHets=hetmet[,2],pop=hetmet$pop,super_pop=hetmet$super_pop)\n",
    "hetmetnice$pop = factor(hetmetnice$pop, levels=unlist(sortedpops))\n",
    "\n",
    "png(\"hets_pop_boxplot.png\",1200,600)\n",
    "boxplot(nHets~pop,data=hetmetnice,\n",
    "    xlab=\"Continental population\",ylab=\"Heterozygous sites\",\n",
    "    col=c(rep(\"blue\",length(sortedpops[[1]])),rep(\"green\",length(sortedpops[[2]])),rep(\"orange\",length(sortedpops[[3]])),rep(\"yellow\",length(sortedpops[[4]])),rep(\"red\",length(sortedpops[[5]]))))\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Now you can see how the heterozygosities differ between super-populations. Boxplots are not an ideal way of presenting data, but convenient in base `R`.\n",
    "One might rather want to get a violin plot with individual dots, which requires more specialized `R` packages: \n",
    "!!check other compi\n",
    "\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "15d62655",
   "metadata": {},
   "source": [
    "# Principle Component Analysis\n",
    "\n",
    "## [What is a PCA?](https://en.wikipedia.org/wiki/Principal_component_analysis)\n",
    "\n",
    "* A statistical method to reduce dimensionality of the data.\n",
    "* *E.g.* 2504 individuals with 2000000 SNPs\n",
    "* Many SNPs are observed across multiple individuals: co-variance\n",
    "* Co-variance shows how similar individuals are - populations cluster with each other\n",
    "* (unsupervised = the method does not know the labels)\n",
    "\n",
    "### PCA in population genetics\n",
    "\n",
    "It is a widely used summary statistic for genotype data. In many papers, it is shown as an exploratory step - we have sequenced new individuals, where do they fall in worldwide or large-scale diversity? This works for humans and many other species with sufficient amounts of data.\n",
    "\n",
    "Here an example from [humans](https://doi.org/10.1038%2Fnature07331): \n",
    "\n",
    "![image.png](attachment:image.png)\n",
    "\n",
    "Here one from [Gorillas](https://doi.org/10.1038%2Fs41559-023-02145-2): \n",
    "\n",
    "![image-2.png](attachment:image-2.png)\n",
    "\n",
    "\n",
    "## A simple PCA in R\n",
    "\n",
    "### Data preparation\n",
    "\n",
    "Now, we have the data on chr18 from the 1000 Genomes data, and we have done quite some things with them. What we need now is a genotype matrix, or a table of numbers which can be used by a statistical method. This table should also not have missing data, or other stuff.\n",
    "\n",
    "Another neat feature of `bcftools` is that it actually can also transform `vcf` formatted files into anything you need. In this case, we really just want to have these numbers without anything else. `bcftools query` is good for that, since you can define which fields to extract, and how to separate them.\n",
    "\n",
    "* Let's do it in one go!\n",
    "\n",
    "```\n",
    "# in BASH\n",
    "\n",
    "bcftools view -m 2 -M 2 -v snps chr18.vcf.gz | bcftools query -f '[%GT ]\\n' > allgts.txt\n",
    "```\n",
    "\n",
    "* This is still a very big file! Let's \"thin\" it a bit by randomly taking 100000 SNPs:\n",
    "\n",
    "```\n",
    "shuf -n 100000 allgts.txt > somegts.txt\n",
    "```\n",
    "\n",
    "There could be different ways of doing this, but here this way is convenient to keep it simple.\n",
    "\n",
    "* Then we can switch into `R`.\n",
    "\n",
    "\n",
    "### Data preparation in R\n",
    "\n",
    "First, we may need the proper library to calculate this! A possible (and very nice) one is [adegenet](https://adegenet.r-forge.r-project.org/). `R` has many packages, and often they are not pre-installed on a computer/structure. If that is the case, you may need to install a package with:\n",
    "\n",
    "```\n",
    "R --vanilla\n",
    "\n",
    "install.packages(\"adegenet\")\n",
    "\n",
    "library(adegenet)\n",
    "\n",
    "```\n",
    "\n",
    "Now, read the SNPS and have a look!\n",
    "\n",
    "```\n",
    "snps<-read.table(\"somegts.txt\", sep=\" \",header=F,comment.char=\"\")\n",
    "head(snps)\n",
    "```\n",
    "\n",
    "* Note that the separator is now a space, because that is how we defined it in the previous step. However, this was adding a last space to the end of the file, and R believes there is a (empty) column.\n",
    "\n",
    "```\n",
    "snps<-snps[,-2505]\n",
    "\n",
    "table(snps[,1])\n",
    "```\n",
    "\n",
    "Now, these are not really numbers, but genotypes...\n",
    "\n",
    "* A very easy way to turn them into single numbers would be using the `ifelse` statement like this:\n",
    "\n",
    "```\n",
    "snps[,1]<-ifelse(snps[,1]==\"0|0\",0,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]==\"1|1\",2,snps[,1])\n",
    "snps[,1]<-ifelse(snps[,1]%in%c(\"0|1\",\"1|0\"),1,snps[,1])\n",
    "table(snps[,1])\n",
    "```\n",
    "\n",
    "* However:\n",
    "\n",
    "```\n",
    "is.numeric(snps[,1])\n",
    "snps[,1]<-as.numeric(snps[,1])\n",
    "is.numeric(snps[,1])\n",
    "```\n",
    "\n",
    "This looks better, but you certainly don't want to type in 2503 more times the same things, and you would very likely make mistakes on the way.\n",
    "\n",
    "* Let's use a `for` loop and nested `ifelse` statements to make it easier! As a side-effect, the whole column will be numeric as well!\n",
    "\n",
    "```\n",
    "for (j in (2:2504)) { snps[,j]<-ifelse(snps[,j]==\"0|0\",0,ifelse(snps[,j]==\"1|1\",2,ifelse(snps[,j]%in%c(\"0|1\",\"1|0\"),1,NA))); print(j) }\n",
    "\n",
    "\n",
    "table(snps[,3])\n",
    "table(snps[,666])\n",
    "is.numeric(snps[,2024])\n",
    "```\n",
    "\n",
    "I would note that at this stage it is ok to do things that way. But next level would be to write a `function` to make it more efficient...\n",
    "\n",
    "If you feel that coming up with efficient solutions is complicated, that is fine - it is a matter of practice and experience. We need to learn things in order to do them.\n",
    "\n",
    "\n",
    "### Calculating the PCA\n",
    "\n",
    "The method we want to use is called `dudi.pca`. It needs the individuals in the rows, and the positions in the columns, that is, a transposed table to be done with `t()`. As you see, nested functions are very common in `R`, do combine things on the fly without intermediate objects. \n",
    "\n",
    "* Let's to this:\n",
    "\n",
    "```\n",
    "pca_object <- dudi.pca(t(snps),nf=20,scannf=F)\n",
    "\n",
    "tb<-list()\n",
    "for (j in (1:ncol(snps)))  { tb[[j]]<-table(snps[,j]) }\n",
    "\n",
    "pca_object\n",
    "summary(pca_object)\n",
    "```\n",
    "\n",
    "This works so far! We have created a higher-level `R` object with lots of information. Important are the coordinates for each individual, which are in the sub-object called `li`, which we can access with the `$` sign.\n",
    "\n",
    "### Plotting the PCA\n",
    "\n",
    "* Let's plot this!\n",
    "\n",
    "```\n",
    "pdf(\"PCA_1000g.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2])\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Ok, this is just some distribution of dots... It is nicer to color it. For this, we need the metadata again. In `R`, we can then create a matching colour vector:\n",
    "\n",
    "```\n",
    "meta<-read.table(\"~/test/integrated_call_samples_v3.20130502.ALL.panel\", sep=\"\\t\",header=T,fill=T)\n",
    "mypops<-c(\"AFR\",\"SAS\",\"EUR\",\"EAS\",\"AMR\")\n",
    "cols=c(\"blue\",\"green\",\"orange\",\"yellow\",\"red\")\n",
    "\n",
    "colvector<-meta$super_pop\n",
    "for (pop in (1:length(mypops))) { colvector<-gsub(mypops[pop],cols[pop],colvector) }\n",
    "```\n",
    "\n",
    "* Let's plot it!\n",
    "\n",
    "```\n",
    "pdf(\"PCA_1000g_col.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector)\n",
    "legend(\"topright\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "This is actually already quite good - considering that it is just a part of chromosome 18! We can do some more polishing to make it look nicer, and add the information how much of the variation is explained by these first two PCs.\n",
    "\n",
    "* Let's add nice axis labels (`xlab` and `ylab`), and points instead of empty circles (with `pch`):\n",
    "\n",
    "\n",
    "```\n",
    "percent_variation=round(pca_object$eig/sum(pca_object$eig)*100,2)\n",
    "pdf(\"PCA_1000g_nice.pdf\",12,12)\n",
    "plot(pca_object$li[,1],pca_object$li[,2],col=colvector,pch=16,\n",
    "    xlab=paste(\"PC1\",\" (\",percent_variation[1],\"%)\",sep=\"\"),\n",
    "    ylab=paste(\"PC\",2,\" (\",percent_variation[2],\"%)\",sep=\"\"),\n",
    "    main=paste(\"PCA of 1000G, chr18\"))\n",
    "legend(\"topright\",legend=mypops, fill=cols,bty=\"n\" )\n",
    "dev.off()\n",
    "```\n",
    "\n",
    "Now, this looks good, and you can gain some insight into the relationships of these populations!\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba73ab4e",
   "metadata": {},
   "source": [
    "# Next step: admixtools!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5cb5045",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
